File: Algorithms/Precompute/Prefix Sums/productExceptSelf.py
from itertools import accumulate
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1]+list(accumulate(nums[:-1], lambda a,b: a*b))
        suffix = list(accumulate(nums[::-1][:-1], lambda a,b: a*b))[::-1] + [1]
        return [a[0]*a[1] for a in zip(prefix, suffix)]

File: Algorithms/Precompute/Prefix Sums/productExceptSelf.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> left(n+1, 1), right(n+1, 1), out(n,1);
        std::partial_sum(nums.begin(), nums.end(), left.begin()+1,std::multiplies<>());
        std::partial_sum(nums.rbegin(), nums.rend(), right.rbegin()+1, std::multiplies<>());
        // std::transform(left.begin(), left.end()-1, right.begin()+1, out.begin(), std::multiplies<>());

        /* right: 24,12,4,1,1 
         * left: 1,1,2,6,24 
         */
        for(int i = 0; i < n; i++)
            out[i] = left[i] * right[i+1];

        return out;
    }
};